lib/minitest/snapshots/serializer.rb

```diff
@@ -0,0 +1,68 @@
+require 'set'
+require 'yaml'
+
+module Minitest
+  module Snapshots
+    # The serializer translates values (objects) into strings. By default,
+    # it uses YAML (Psych). Can be overridden to implement custom
+    # serialization formats, e.g.:
+    #
+    #  module Minitest::Snapshots::Serializer
+    #    def self.serialize(value)
+    #      Marshal.dump(value)
+    #    end
+    #  end
+    module Serializer
+      # The name of the method used to customize the output of +to_yaml+.
+      # Used to provide canonical representations for Hash and Set instances.
+      HOOK = :encode_with
+
+      # Used to ensure the addition/removal of hooks is atomic
+      @lock = Mutex.new
+
+      # Serialize the supplied value to YAML with hooks to canonicalize
+      # (i.e. sort) Hash keys and Set elements.
+      #
+      #   h1 = { foo: "bar", baz: "quux" }
+      #   h2 = { baz: "quux", foo: "bar" }
+      #
+      #   h1.to_yaml == h2.to_yaml                             # => false
+      #   Serializer.serialize(h1) == Serializer.serialize(h2) # => true
+      #
+      # The hooks are only installed for the duration of the method call
+      # and are not installed if custom hooks are already defined.
+      def self.serialize(value)
+        @lock.synchronize do
```

Comment:
🤦 I understand why you had to do this but this seems like a lot of complexity coming from the fact that Pysch is missing a way to canonicalize the keys itself. It also adds some overhead and concurrency limitations that would be otherwise unnecessary if we used a different serializer. It's a clean way of making this all work without monkey-patching. 

What do you think about using something like https://github.com/Phrogz/NeatJSON instead which just supports sorted keys already and keeps it simple? I feel like YAML's ability to dump full objects marshal-style is kind of nice, and switching to JSON would drop that, but that maybe forcing developers writing tests to only pass this library primitives is actually a good thing so that the snapshots aren't super deeply coupled to the internal structure of the code under test.

Reply:
Thanks for your feedback!

I'll have a play with NeatJSON. One thing I'll say for YAML is that it leads to nice readable diffs when there's a mismatch. I suspect that will be lost slightly with JSON.

> I feel like YAML's ability to dump full objects marshal-style is kind of nice, and switching to JSON would drop that, but that maybe forcing developers writing tests to only pass this library primitives is actually a good thing so that the snapshots aren't super deeply coupled to the internal structure of the code under test.

I've never had to do that (simplify my data to satisfy the serializer) when using snapshots with JavaScript (with [AVA](https://github.com/avajs/ava/blob/master/docs/04-snapshot-testing.md)). IMO, these tools are meant to serve us rather than the other way round. If people want that restriction, I think it would be better for them to adopt that policy explicitly rather than for the tool to impose it in an opinionated way, particularly as the rationale feels *ex post facto* to me. After all, if this option was built into Psych, we'd just use that.

> What do you think about using something like https://github.com/Phrogz/NeatJSON instead which just supports sorted keys already and keeps it simple?

I think it makes more sense for this library to implement the most faithful (and IMO least surprising) serializer by default and for users to replace it with a simpler version if they don't need the fidelity, rather than the other way round. People aren't going to implement a canonical YAML serializer themselves (I had to figure it out by guessing since Psych's documentation is so sparse), but the extension point would make it straightforward to use an alternative serializer if that's sufficient:

```ruby
require "neatjson"

module Minitest::Snapshots::Serializer
  def self.serialize(value)
    JSON.pretty_generate(value, sort: true)
  end
end
```

Although, note that it doesn't work with Sets (it just calls `inspect`):

```ruby
s1 = Set.new([1, 2, 3, 4, 5])
s2 = Set.new([5, 4, 3, 2, 1])

JSON.neat_generate(s1, sort_true) # => "\"#<Set: {1, 2, 3, 4, 5}>\""
JSON.neat_generate(s2, sort_true) # => "\"#<Set: {5, 4, 3, 2, 1}>\""
JSON.neat_generate(s1, sort: true) == JSON.neat_generate(s2, sort: true) #=> false
```

Reply:
Ok, I am convinced, better to do backflips inside the gem than make every user of it do poor ones themselves.